src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs

```diff
@@ -0,0 +1,313 @@
+
+            // So, problem is, it could be that the missing root was just not processed, so we can't just not retry
+            // forever. `DateTimeOffset.Now` is probably heavy though.
+            _rootWithMissingNode.Delete(rootHash);
+        }
+
+        return false;
+    }
+
+    private void TrackTrieNodeException(in ValueHash256 rootHash)
+    {
+        _rootWithMissingNode.Set(rootHash, DateTimeOffset.Now);
+    }
+
+    public byte[][]? GetTrieNodes(PathGroup[] pathSet, in ValueHash256 rootHash, CancellationToken cancellationToken)
+    {
+        if (IsRootMissing(rootHash)) return Array.Empty<byte[]>();
+
+        _logger.Debug($"Trie node {pathSet.Length}");
+        // TODO: use cache to reduce node retrieval from disk
+        int pathLength = pathSet.Length;
+        List<byte[]> response = new(pathLength);
+        StateTree tree = new(_store, _logManager);
+        bool abort = false;
+
+        for (int reqi = 0; reqi < pathLength && !abort && !cancellationToken.IsCancellationRequested; reqi++)
+        {
+            byte[][]? requestedPath = pathSet[reqi].Group;
+            switch (requestedPath.Length)
+            {
+                case 0:
+                    return null;
+                case 1:
+                    try
+                    {
+                        byte[]? rlp = tree.GetNodeByPath(Nibbles.CompactToHexEncode(requestedPath[0]), rootHash.ToCommitment());
+                        response.Add(rlp);
+                    }
+                    catch (MissingTrieNodeException)
+                    {
+                        TrackTrieNodeException(rootHash);
+                        abort = true;
+                    }
+                    break;
+                default:
+                    try
+                    {
+                        Account? account = GetAccountByPath(tree, rootHash, requestedPath[0]);
+                        if (account is null)
+                        {
+                            break;
+                        }
+                        Hash256? storageRoot = account.StorageRoot;
+                        if (storageRoot.Bytes.SequenceEqual(Keccak.EmptyTreeHash.Bytes))
+                        {
+                            break;
+                        }
+                        StorageTree sTree = new(_store, storageRoot, _logManager);
+
+                        for (int reqStorage = 1; reqStorage < requestedPath.Length; reqStorage++)
+                        {
+                            byte[]? sRlp = sTree.GetNodeByPath(Nibbles.CompactToHexEncode(requestedPath[reqStorage]));
+                            response.Add(sRlp);
+                        }
+                    }
+                    catch (MissingTrieNodeException)
+                    {
+                        TrackTrieNodeException(rootHash);
+                        abort = true;
+                    }
+                    break;
+            }
+        }
+        return response.ToArray();
```

Comment:
regardless we have List and the ToArray, ArrayPoolList would at least reduce allocations a bit